source/val/validate_decorations.cpp

```diff
@@ -2083,6 +2083,197 @@ spv_result_t CheckDecorationsFromDecoration(ValidationState_t& vstate) {
   return SPV_SUCCESS;
 }
 
+bool AllowsLayout(ValidationState_t& vstate, const spv::StorageClass sc) {
+  switch (sc) {
+    case spv::StorageClass::StorageBuffer:
+    case spv::StorageClass::Uniform:
+    case spv::StorageClass::PhysicalStorageBuffer:
+    case spv::StorageClass::PushConstant:
+      // Always explicitly laid out.
+      return true;
+    case spv::StorageClass::UniformConstant:
+      return false;
+    case spv::StorageClass::Workgroup:
+      return vstate.HasCapability(
+          spv::Capability::WorkgroupMemoryExplicitLayoutKHR);
+    case spv::StorageClass::Function:
+    case spv::StorageClass::Private:
+      return vstate.version() < SPV_SPIRV_VERSION_WORD(1, 4);
```

Comment:
@alan-baker so for something in GLSL like

```
#version 450

struct Bar {
    uint x;
    uint y;
    uint z[2];
};

layout(set = 0, binding = 0, std430) buffer foo {
    uvec4 a;
    Bar b; // size 16 at offset 16
    uint c;
};

void main() {
    Bar new_bar;
    b = new_bar;
}
```

this will pass `spirv-val` if using `--target-env vulkan1.1` (https://godbolt.org/z/bjKodYPKG)

but fail here with `1.2`

I guess trying to figure out what spec change causes this to fail from 1.1 to 1.2

Reply:
> This chapter describes valid uses for a set of SPIR-V decorations. Any other use of one of these decorations is invalid, with the exception that, when using SPIR-V versions 1.4 and earlier: Block, BufferBlock, Offset, ArrayStride, and MatrixStride can also decorate types and type members used by variables in the Private and Function storage classes.

Vulkan 1.2 is considered SPIR-V 1.5 (for the assembler) so that prevents putting layouts on types used in function variables.

Looks like I misread the spec though and should change the conditions to <= 1.4.

Reply:
Ok, this was in the `Shader Interfaces` section of the Vulkan Spec (I was searching only in the SPIR-V spec)

Reply:
That fix is #6023